crates/hir_ty/src/diagnostics/expr.rs

```diff
@@ -150,20 +163,58 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
         }
     }
 
-    fn validate_call(&mut self, db: &dyn HirDatabase, call_id: ExprId, expr: &Expr) -> Option<()> {
+    fn check_for_filter_map_next(&mut self, db: &dyn HirDatabase) {
+        let body = db.body(self.owner.into());
+        let mut prev = None;
+
+        for (id, expr) in body.exprs.iter() {
+            if let Expr::MethodCall { receiver, method_name, args, .. } = expr {
+                let method_name_hack_do_not_merge = format!("{}", method_name);
+
+                if method_name_hack_do_not_merge == "filter_map" && args.len() == 1 {
+                    prev = Some((id, args[0]));
+                    continue;
+                }
+
+                if method_name_hack_do_not_merge == "next" {
+                    if let Some((filter_map_id, filter_map_args)) = prev {
+                        if *receiver == filter_map_id {
+                            let (_, source_map) = db.body_with_source_map(self.owner.into());
+                            if let (Ok(filter_map_source_ptr), Ok(next_source_ptr)) = (
```

Comment:
Using the example function bar() above, the range of next_expr is

```
[1,` 2, 3].iter() .filter_map(|x| if *x == 2 { Some(4) } else { None }).next()
```

and the range of filter_map_expr is

```
[1, 2, 3].iter() .filter_map(|x| if *x == 2 { Some(4) } else { None })
```

Should I just walk down the syntax tree here to find the bits I want to replace (METHOD_CALL_EXPR -> METHOD_CALL_EXPR -> NAME_REF...) or is there a better way of doing this?